core/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java

```diff
@@ -0,0 +1,257 @@
+
+        private final ActionListener<RolloverResponse> listener;
+        private final String oldIndex;
+        private final String newIndex;
+
+        public IndicesAliasesListener(String oldIndex, String newIndex, ActionListener<RolloverResponse> listener) {
+            this.oldIndex = oldIndex;
+            this.newIndex = newIndex;
+            this.listener = listener;
+        }
+
+        @Override
+        public void onResponse(ClusterStateUpdateResponse clusterStateUpdateResponse) {
+            listener.onResponse(new RolloverResponse(oldIndex, newIndex));
+        }
+
+        @Override
+        public void onFailure(Throwable e) {
+            listener.onFailure(e);
+        }
+    }
+
+    static IndicesAliasesClusterStateUpdateRequest prepareIndicesAliasesRequest(String concreteSourceIndex, String targetIndex,
+                                                                                RolloverRequest request) {
+        final IndicesAliasesClusterStateUpdateRequest updateRequest = new IndicesAliasesClusterStateUpdateRequest()
+            .ackTimeout(request.ackTimeout())
+            .masterNodeTimeout(request.masterNodeTimeout());
+        final AliasAction[] actions;
+        if (request.getOptionalTargetAlias() != null) {
+            actions = new AliasAction[3];
+            actions[0] = new AliasAction(AliasAction.Type.ADD, targetIndex, request.getSourceAlias());
+            actions[1] = new AliasAction(AliasAction.Type.ADD, targetIndex, request.getOptionalTargetAlias());
+            actions[2] = new AliasAction(AliasAction.Type.REMOVE, concreteSourceIndex, request.getSourceAlias());
+        } else {
+            actions = new AliasAction[2];
+            actions[0] = new AliasAction(AliasAction.Type.ADD, targetIndex, request.getSourceAlias());
+            actions[1] = new AliasAction(AliasAction.Type.REMOVE, concreteSourceIndex, request.getSourceAlias());
+        }
+        updateRequest.actions(actions);
+        return updateRequest;
+    }
+
+    static String generateRolloverIndexName(String sourceIndexName) {
+        int numberIndex = sourceIndexName.lastIndexOf("-");
+        int counter = 1;
+        String indexPrefix = sourceIndexName;
+        if (numberIndex != -1) {
+            try {
+                counter = Integer.parseInt(sourceIndexName.substring(numberIndex + 1));
+                counter++;
+                indexPrefix = sourceIndexName.substring(0, numberIndex);
+            } catch (NumberFormatException ignored) {
+            }
+        }
+        return String.join("-", indexPrefix, String.valueOf(counter));
+    }
+
+    static boolean satisfiesConditions(List<Condition> requestConditions, long docCount, long sizeInBytes,
+                                       long creationDate) {
+        for (Condition condition: requestConditions) {
+            switch (condition.getType()) {
+                case MAX_SIZE:
```

Comment:
This feels like it maybe should be an abstract method in ConditionType and each type overrides?